sqldb/invoices.go

```diff
@@ -0,0 +1,1554 @@
+			CltvDelta: sqlInt32(newInvoice.Terms.FinalCltvDelta),
+			Expiry:    int32(newInvoice.Terms.Expiry),
+			// Note: keysend invoices don't have a payment request.
+			PaymentRequest: sqlStr(string(
+				newInvoice.PaymentRequest),
+			),
+			State:          int16(newInvoice.State),
+			AmountPaidMsat: int64(newInvoice.AmtPaid),
+			IsAmp:          newInvoice.IsAMP(),
+			IsHodl:         newInvoice.HodlInvoice,
+			IsKeysend:      newInvoice.IsKeysend(),
+			CreatedAt:      newInvoice.CreationDate.UTC(),
+		}
+
+		// Some invoices may not have a preimage, like in the case of
+		// HODL invoices.
+		if newInvoice.Terms.PaymentPreimage != nil {
+			preimage := *newInvoice.Terms.PaymentPreimage
+			if preimage == invpkg.UnknownPreimage {
+				return errors.New("cannot use all-zeroes " +
+					"preimage")
+			}
+			params.Preimage = preimage[:]
+		}
+
+		// Some non MPP payments may have the defaulf (invalid) value.
+		if newInvoice.Terms.PaymentAddr != invpkg.BlankPayAddr {
+			params.PaymentAddr = newInvoice.Terms.PaymentAddr[:]
+		}
+
+		var err error
+		invoiceID, err = db.InsertInvoice(ctx, params)
+		if err != nil {
+			return fmt.Errorf("unable to insert invoice: %w", err)
+		}
+
+		// TODO(positiveblue): if invocies do not have custom features
+		// maybe just store the "invoice type" and populate the features
+		// based on that.
+		for feature := range newInvoice.Terms.Features.Features() {
+			params := sqlc.InsertInvoiceFeatureParams{
+				InvoiceID: invoiceID,
+				Feature:   int32(feature),
+			}
+
+			err := db.InsertInvoiceFeature(ctx, params)
+			if err != nil {
+				return fmt.Errorf("unable to insert invoice "+
+					"feature(%v): %w", feature, err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		// Add context to unique constraint errors.
+		var uniqueConstraintErr *ErrSQLUniqueConstraintViolation
+		if errors.As(err, &uniqueConstraintErr) {
+			return 0, invpkg.ErrDuplicateInvoice
+		}
+
+		return 0, fmt.Errorf("unable to add invoice(%v): %w",
+			paymentHash, err)
+	}
+
+	newInvoice.AddIndex = uint64(invoiceID)
+
+	return newInvoice.AddIndex, nil
+}
+
+// fetchInvoiceData fetches the common invoice data and the AMP state for the
+// invoice with the given reference.
+func (i *InvoiceStore) fetchInvoice(ctx context.Context,
+	db InvoiceQueries, ref invpkg.InvoiceRef) (*invpkg.Invoice, error) {
+
+	if ref.PayHash() == nil && ref.PayAddr() == nil && ref.SetID() == nil {
+		return nil, invpkg.ErrInvoiceNotFound
+	}
+
+	var (
+		invoice *invpkg.Invoice
+		params  sqlc.GetInvoiceParams
+	)
+
+	if ref.PayHash() != nil {
+		params.Hash = ref.PayHash()[:]
+	}
+
+	if ref.PayAddr() != nil && *ref.PayAddr() != invpkg.BlankPayAddr {
+		params.PaymentAddr = ref.PayAddr()[:]
+	}
+
+	// Fetch all AMP HTLCs, unless it is explicitly requested to only fetch
+	// the AMP HTLCs for this specific set.
+	if ref.SetID() != nil && ref.Modifier() != invpkg.HtlcSetOnlyModifier {
+		params.SetID = ref.SetID()[:]
```

Comment:
So I don't think we care about this param here. We're just fetching an invoice, not the payments to the invoice, so the `set_id` value won't be used.